### PR TITLE
Bugfix: unlink global file not found locally

### PR DESCRIPTION
**Pull request by CamStan**
### Description

Add logic to unlink/remove a file that exists globally when it is not found locally.

Add option to reuse a file name, and a function to remove a file to _testutil.h_.
Add check for reusing a file name to _write_, _writeread_, and _cr_ examples.

Change default `UNIFYFS_LOGIO_SPILL_SIZE` to 4GIB.
Closes: #634 

Minor documentation clarifications.

### Motivation and Context

Idea was brought up to allow rerunning the examples multiple times with the same filename.

Attempting this resulted in being able to stat a file to show it existed, but when attempting to remove a file created by a different client, unifyfs would claim the file didn't exist.

### How Has This Been Tested?

Ran _write-gotcha/static_ example with and without the new `-r|--reuse-filename` option.
E.g.,
```
jsrun -r1 -n2 -a1 -c1 -e individual -k $UNIFYFS_LOG_DIR/write-gotcha.err -o $UNIFYFS_LOG_DIR/write-gotcha.out $UNIFYFS_EXAMPLES/write-gotcha -v -r -f mytestfile

jsrun -r1 -n2 -a1 -c1 -e individual -k $UNIFYFS_LOG_DIR/write-gotcha-reuse.err -o $UNIFYFS_LOG_DIR/write-gotcha-reuse.out $UNIFYFS_EXAMPLES/write-gotcha -v -r -f mytestfile
```

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Testing (addition of new tests or update to current tests)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the UnifyFS code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted.

### CODE REVIEW

File: client/src/unifyfs-sysio.c [hunk 1]
```diff
@@ -616,27 +638,51 @@ int UNIFYFS_WRAP(remove)(const char* path)
...
-            errno = unifyfs_rc_errno(ret);
-            return -1;
+            /* check that it's not a directory */
+            if (attr.mode & S_IFDIR) {
+                /* TODO: shall be equivalent to rmdir(path) */
+                /* ERROR: is a directory */
+                LOGDBG("Attempting to unlink a directory %s in UNIFYFS", upath);
+                errno = EISDIR;
+                return -1;
+            }
```
Comment (adammoody) [inline client/src/unifyfs-sysio.c:660]: I think it's valid to call ``remove`` on a directory.  I see we also had this check for a directory in our ``remove`` wrapper before this change.
```
NAME
       remove - remove a file or directory

SYNOPSIS
       #include <stdio.h>

       int remove(const char *pathname);

DESCRIPTION
       remove() deletes a name from the file system.  It calls unlink(2) for files, and rmdir(2) for directories.
```

**Review by adammoody — APPROVED** [review]

**Comment by adammoody** [comment]
Thanks @CamStan .  Nice work tracking this down and finding the fix.  This looks good to me, other than the remove-on-directory item.

BTW, our local file cache can get stale at times.  That requires a bigger fix than what we'd want to do here, but this provides a good opportunity to highlight a problem.  For example, I think this related 2-process test won't work like we'd want it to.
```
char[] path = "/unifyfs/foo"
if (rank == 1) {
  mkdir(path) // local cache on rank 1 will record path as a direcotry
}

MPI_Barrier(MPI_COMM_WORLD)

if (rank == 0) {
  rmdir(path) // rank 0 deletes item
  int fd = open(path, O_WRONLY | O_CREAT) // recreates it as a file
  close(fd)
}

MPI_Barrier(MPI_COMM_WORLD)

if (rank == 1) {
  unlink(path) // cache on rank 1 still sees item as directory, even though it's now actually a file, so unlink will fail
}
```

File: client/src/unifyfs-sysio.c (hunk 1 above)
Comment (CamStan) [reply client/src/unifyfs-sysio.c:660]: Updated this for when the file is found locally with the following. However, haven't added anything to address this when the dir is found globally and this brought up a few questions. I'll explain this in my comment below.
``` C
         } else { /* got the file locally */
            /* if directory, is it empty? */
            if (unifyfs_fid_is_dir(fid) && !unifyfs_fid_is_dir_empty(upath)) {
                /* ERROR: is a directory, but isn't empty */
                LOGDBG("Attempting to remove non-empty dir %s in UNIFYFS",
                        upath);
                errno = ENOTEMPTY;
                 return -1;
             }
```

**Comment by CamStan** [comment]
@adammoody, I went to update this for having `remove()` behave like `rmdir()` and noticed there may be a few other bugs here.

1. The `rmdir()` wrapper has the same issue that this PR attempts to fix with `unlink()` and `remove()`, in that the existence of the directory is only checked for locally.

2. The `unifyfs_fid_is_dir_empty()` function only checks that the directory is local and in use.
https://github.com/LLNL/UnifyFS/blob/2a0352a3f2a66959157ae828cb16412186f182d3/client/src/unifyfs.c#L653-L658
Would we want to update this to also check globally, write a different function, or just assume that if the directory is only found globally, that it's not in use and is okay to remove (this is assuming 3 is fine as it is)?

3. Realized that our unit test for removing a non-empty dir fails, despite the `rmdir()` wrapper checking for this:
https://github.com/LLNL/UnifyFS/blob/2a0352a3f2a66959157ae828cb16412186f182d3/client/src/unifyfs-sysio.c#L136-L140
I'm assuming this is because of the `in_use` check mentioned in 2. However, if this is the behavior we want, then the solution to just let `remove()`work on a directory only found globally is easy.
```
    PASS: 0100-sysio-gotcha.t 20 - ../../../t/sys/mkdir-rmdir.c:125 creat subfile /tmp/tmp.S9GVtl4fS6/mount/PTTcS._fpC4Z4ztupby0s4@BOxRm+FoOYAyy./subfile in existing dir (fd=30000): Success
    XFAIL: 0100-sysio-gotcha.t 26 - ../../../t/sys/mkdir-rmdir.c:174 rmdir non-empty directory /tmp/tmp.S9GVtl4fS6/mount/PTTcS._fpC4Z4ztupby0s4@BOxRm+FoOYAyy. should fail (errno=0): Success # TODO mkdir_5: unifyfs currently creates all paths as separate entities
```
I can create an issue for these (and the one two-process one you mentioned) so we don't lose them, unless we want to try to address some/all in this PR.

**Comment by CamStan** [comment]
This is ready for another review.
Checkpatch is failing because of my middle commit that I've only kept for reference. If we want to use this approach, and when this is ready to merge, I'll squash these.

**Review by MichaelBrim — COMMENTED** [review]
I see a couple places in where you check the shared logic function's mask against S_IFLNK, but no calls to the function set that flag in the mask.

**Comment by CamStan** [comment]
> I see a couple places in where you check the shared logic function's mask against S_IFLNK, but no calls to the function set that flag in the mask.

True. Checking for this came up in one of my discussions with Adam, but was before I had thought through the logic.

I just noticed we don't seem to wrap `symlink()` and at first thought, could UnifyFS even have symlinks in it, the way it's currently set up?

I'm not treating them any different than regular files here as the only extra thing our current removes wrappers really care about right now is if something is a directory or not.
Should probably just remove that check then, yeah? Then if we end up supporting them, anything special about their removal can be added.

**Comment by MichaelBrim** [comment]
I don't think we should worry about symlink support at this time. I can't think of a practical use case where an HPC app would need to use a symlink. And you're right that supporting them using our current codebase would actually be quite difficult.

File: client/src/unifyfs-sysio.c [hunk 2]
```diff
@@ -109,42 +109,124 @@ int UNIFYFS_WRAP(mkdir)(const char* path, mode_t mode)
...
 }
 
-int UNIFYFS_WRAP(rmdir)(const char* path)
+/* Function for shared logic between rmdir, remove, and unlink wrappers.
+ *
+ * Remove the target at upath based on the input file mode indicated by mask.
+ * If mask == 0 (e.g., remove wrapper made request), the target doesn't need
+ * any mode specific checks before attempting to remove.
+ * */
+static int unifyfs_remove_mask(const char* upath, mode_t mask)
```
Comment (MichaelBrim) [inline client/src/unifyfs-sysio.c:118]: For code readability, I personally think this function should be named differently. If you just read the name, you would think it "removes a mask" from the file, which it doesn't. Perhaps just call it "unifyfs_remove"? The function comment clearly states the purpose of the mask parameter.

Comment (CamStan) [reply client/src/unifyfs-sysio.c]: Good call. Changed this.

**Review by MichaelBrim — APPROVED** [review]

**Pull request merged**